internal/usde/usde.go

```diff
@@ -1,39 +1,103 @@
+		// TODO-PROGRESSIVE: return UpgradeStrategyProgressive instead of UpgradeStrategyPPND
+		return UpgradeStrategyPPND, nil, nil
+	}
+
+	// Get USDE Config
+	usdeConfig := config.GetConfigManagerInstance().GetUSDEConfig()
+
+	// Get apply paths based on the spec type (Pipeline, ISBS)
+	applyPaths := []string{}
+	if reflect.DeepEqual(newSpec.GroupVersionKind(), numaflowv1.PipelineGroupVersionKind) {
+		applyPaths = usdeConfig.PipelineSpecExcludedPaths
+	} else if reflect.DeepEqual(newSpec.GroupVersionKind(), numaflowv1.ISBGroupVersionKind) {
+		applyPaths = usdeConfig.ISBServiceSpecExcludedPaths
+	}
+
+	numaLogger.WithValues("usdeConfig", usdeConfig, "applyPaths", applyPaths).Debug("started deriving upgrade strategy")
+
+	// Split newSpec
+	newSpecOnlyApplyPaths, newSpecWithoutApplyPaths, err := util.SplitObject(newSpec.Spec.Raw, applyPaths, ".")
+	if err != nil {
+		return UpgradeStrategyError, nil, err
+	}
+
+	numaLogger.WithValues("newSpecOnlyApplyPaths", newSpecOnlyApplyPaths, "newSpecWithoutApplyPaths", newSpecWithoutApplyPaths).Debug("split new spec")
 
-func (s *USDEStrategy) UnmarshalJSON(data []byte) (err error) {
-	// Trim spaces and check length
-	dataStrNoSpaces := strings.TrimSpace(string(data[:]))
-	if len(dataStrNoSpaces) == 0 {
-		return fmt.Errorf("empty strategy (allowed values are: %s or %s)", progressiveStrategyID, ppndStrategyID)
+	// Split existingSpec
+	existingSpecOnlyApplyPaths, existingSpecWithoutApplyPaths, err := util.SplitObject(existingSpec.Spec.Raw, applyPaths, ".")
+	if err != nil {
+		return UpgradeStrategyError, nil, err
 	}
 
-	// Remove the double quotes around the string
-	strategyStr := string(data[1 : len(data)-1])
+	numaLogger.WithValues("existingSpecOnlyApplyPaths", existingSpecOnlyApplyPaths, "existingSpecWithoutApplyPaths", existingSpecWithoutApplyPaths).Debug("split existing spec")
+
+	specsDiffer := true
+
+	// Compare specs without the apply fields and check user's strategy to either return PPND or Progressive
+	if !reflect.DeepEqual(newSpecWithoutApplyPaths, existingSpecWithoutApplyPaths) {
+		namespaceConfig := config.GetConfigManagerInstance().GetNamespaceConfig(newSpec.Namespace)
+
+		var userUpgradeStrategy config.USDEUserStrategy = ""
+		if namespaceConfig != nil {
+			userUpgradeStrategy = namespaceConfig.UpgradeStrategy
+		}
+
+		numaLogger.WithValues("userUpgradeStrategy", userUpgradeStrategy).Debug("the specs without the 'apply' paths are different")
+
+		if userUpgradeStrategy == config.PPNDStrategyID {
+			return UpgradeStrategyPPND, &specsDiffer, nil
+		}
 
-	// Make sure the string is one of the possible strategy values
-	if strategyStr != progressiveStrategyID && strategyStr != ppndStrategyID {
-		return fmt.Errorf("invalid strategy %s (allowed values are: %s or %s)", string(data[:]), progressiveStrategyID, ppndStrategyID)
+		// TODO-PROGRESSIVE: return UpgradeStrategyProgressive instead of UpgradeStrategyPPND
+		return UpgradeStrategyPPND, &specsDiffer, nil
```

Comment:
am I correct that right now we're only doing PPND? so, basic CRUD is not an option for users?

Reply:
I believe this is what we discussed with Vigith - just want to confirm it

Reply:
If I recall correctly, that is what we decided and we will use the feature flag to control enabling/disabling data loss prevention via PPND.